### DISCUSSION TITLE
Performance with web bundle

### DISCUSSION THREAD

**Discussion by rothsandro** [shikijs/shiki#846] [General]
Hi,

I'm migrating my [Eleventy template](https://github.com/rothsandro/eleventy-notes/pull/64) from Prism (via [Eleventy Syntax Highlight Plugin](https://www.11ty.dev/docs/plugins/syntaxhighlight/)) from Shiki using the markdown-it plugin.

```js
import { fromHighlighter } from "@shikijs/markdown-it/core";
import { createHighlighter, bundledLanguages } from "shiki/bundle/web";

const highlighter = await createHighlighter({
  langs: Object.keys(bundledLanguages),
  themes: ["light-plus", "dark-plus"],
});

markdownIt.use(
  fromHighlighter(highlighter, {
    defaultLanguage: "text",
    fallbackLanguage: "text",
    themes: {
      light: "light-plus",
      dark: "dark-plus",
    },
  })
)
```

For testing purposes I took an existing Markdown file from the repo, duplicated it 1200 times and then run the Eleventy build:

| Library                 | Build Time | Shiki Setup |
|-------------------------|-----------:|------------:|
| Prism                   | 10s        | n/a         |
| Shiki (with web bundle) | 46s        | 760ms       |
| Shiki (with 8 langs)    | 14s        | 45ms        |

The Shiki Setup (`createHighlighter` + `fromHighlighter`) takes longer the more languages we initialize - which is expected and also acceptable as this happens only once and does not depend on the number of Markdown files.

Shiki with 8 langs is slower than Prism (10s -> 14s) but still acceptable. But with the web bundle initialized the build is way slower (14s -> 46s). Is this really expected? I thought that  `createHighlighter()` already loads the languages (that's why it is slower) and running syntax highlighting should not affect how many languages are loaded (as the number of languages actually used in Markdown files is the same).

**Comment by antfu** [comment]
Thanks for bringing this up, I updated the docs with some guides: https://github.com/shikijs/shiki/commit/5ef5d56c71d5748e65881e4bec9a6a12a6f577a5

https://shiki.style/guide/best-performance

**Reply by rothsandro** [reply]
Thanks for the additional docs.

- **Cache the Highlighter Instance**: I already do this, `createHighlighter()` is only called once.
- **Fine-Grained Bundle**: I only load two themes and the web bundle for languages (but I would actually need to to load all languages as I don't know which languages the user needs). 
- **Use Shorthands**: Unfortunately not possible with `markdown-it` :(
- **JavaScript Engine and Pre-compiled Languages**: I tested the JS Engine and it's slower with just a few Markdown files but faster with 1200 files. It's still way slower than Prism. The experimental `@shikijs/langs-precompiled` package doesn't make much difference in my tests.
- **Use Workers**: I'll test it when docs are ready.

**Reply by antfu** [reply]
For `markdown-it`: I made https://github.com/antfu/markdown-it-async, you could give a try (but that would requires manual integrations for now).

> It's still way slower than Prism

It's unfortunately not the goal of Shiki. Shiki is way more complex than Prism and the target is to provide the most accurate highlighting result. While we aim to be as performant as we can, being faster than Prism is hard, if ever possible.

**Reply by rothsandro** [reply]
I tried `markdown-it-async` and it's much faster (~15s), still 50% slower than Prism but that's as you mentioned that's expected.  Integrating this into Eleventy is not that straight forward. I just overwrite the `render()` function, call `renderAsync()` and return a promise. Eleventy does not officially support this but it still works (because it's the last thing they call in a async function). This could easily break in a future version if they change something internally.

Thanks a lot for your help 🙏
